scripts/run_e2e_tests.py

```diff
@@ -222,397 +117,172 @@ def is_oppia_server_already_running():
-            if len(nextline) == 0 and p.poll() is not None:
-                break
-            if isinstance(nextline, str):
-                # This is a failsafe line in case we get non-unicode input,
-                # but the tests provide all strings as unicode.
-                nextline = nextline.decode('utf-8')  # pragma: nocover
-            output_lines.append(nextline.rstrip())
+        # Keep reading lines until an empty string is returned. Empty strings
+        # signal that the process has ended.
+        for line in iter(proc.stdout.readline, b''):
+            if isinstance(line, str):
+                # Although our unit tests always provide unicode strings, the
+                # actual server needs this failsafe since it can output
+                # non-unicode strings.
+                line = line.decode('utf-8') # pragma: nocover
+            output_lines.append(line.rstrip())
             # Replaces non-ASCII characters with '?'.
-            sys.stdout.write(nextline.encode('ascii', errors='replace'))
+            sys.stdout.write(line.encode('ascii', errors='replace'))
 
-        return output_lines, p.returncode
+        return output_lines, proc.returncode
 
 
 def main(args=None):
     """Run tests, rerunning at most MAX_RETRY_COUNT times if they flake."""
     parsed_args = _PARSER.parse_args(args=args)
 
-    portserver_process = start_portserver()
-    atexit.register(cleanup_portserver, portserver_process)
+    with common.managed_portserver():
+        for attempt_num in python_utils.RANGE(1, MAX_RETRY_COUNT + 1):
+            python_utils.PRINT('***Attempt %d.***' % attempt_num)
 
-    for attempt_num in python_utils.RANGE(MAX_RETRY_COUNT):
-        python_utils.PRINT('***Attempt %s.***' % (attempt_num + 1))
-        output, return_code = run_tests(parsed_args)
-        # Don't rerun off of CI.
-        if not flake_checker.check_if_on_ci():
-            python_utils.PRINT('No reruns because not running on CI.')
-            break
-        # Don't rerun passing tests.
-        if return_code == 0:
-            flake_checker.report_pass(parsed_args.suite)
-            break
-        flaky = flake_checker.is_test_output_flaky(
-            output, parsed_args.suite)
-        # Don't rerun if the test was non-flaky and we are not
-        # rerunning non-flaky tests.
-        if not flaky and not RERUN_NON_FLAKY:
-            break
-        # Prepare for rerun.
-        cleanup()
+            output, return_code = run_tests(parsed_args)
+
+            if not flake_checker.check_if_on_ci():
+                # Don't rerun off of CI.
+                python_utils.PRINT('No reruns because not running on CI.')
+                break
+
+            if return_code == 0:
+                # Don't rerun passing tests.
+                flake_checker.report_pass(parsed_args.suite)
+                break
+
+            test_is_flaky = (
+                flake_checker.is_test_output_flaky(output, parsed_args.suite))
```

Comment:
I don't think we need the outer parentheses (though I haven't counted the columns):

```suggestion
            test_is_flaky = flake_checker.is_test_output_flaky(
                output, parsed_args.suite)
```

Reply:
Done.